net/libfab/libfab.c

```diff
@@ -208,37 +209,54 @@ static int libfab_end_point_create(struct m0_net_end_point **epp,
 	return M0_RC(result);
 }
 
+
 /**
- * Initialises a network buffer.
+ * Register a network buffer.
  *
  * Used as m0_net_xprt_ops::xo_buf_register().
  *
  * @see m0_net_buffer_register().
  */
 static int libfab_buf_register(struct m0_net_buffer *nb)
 {
-	int        		ret = 0;
-
+	int			      ret;
+	struct fid_domain	     *domain = nb->nb_dom->nd_xprt_private;
+	struct m0_fab__buf_params    *bp;
+
+	M0_PRE(nb->nb_xprt_private == NULL);
+
+	M0_ALLOC_PTR(bp);
+	if (bp == NULL)
+		return M0_ERR(-ENOMEM);
+
+	nb->nb_xprt_private = bp;
+	bp->fbp_nb = nb;
+	/* Registers buffer that can be used for send/recv and local/remote RMA. */
+	ret = fi_mr_reg(domain, nb->nb_buffer.ov_buf[0], nb->nb_length,
+			FAB_MR_ACCESS, FAB_MR_OFFSET, FAB_MR_KEY,
+			FAB_MR_FLAG, &bp->fbp_mr, NULL);	
+	if (ret == FI_SUCCESS)
+		ret = fi_mr_enable(bp->fbp_mr);	
 	return M0_RC(ret);
-
-	/*
-	* TODO
-	* fi_mr_reg, fi_mr_desc, fi_mr_key, fi_mr_bind, fi_mr_enable
-	*/
-}	
+}
 
 /**
- * Finalises a network buffer.
+ * Deregister a network buffer.
  *
  * Used as m0_net_xprt_ops::xo_buf_deregister().
  *
  * @see m0_net_buffer_deregister().
  */
 static void libfab_buf_deregister(struct m0_net_buffer *nb)
 {
-	/*
- 	* TODO:
- 	* fi_close() */
+	struct m0_fab__buf_params *bp = nb->nb_xprt_private;
+	int			   ret;
+
+	ret = fi_close(&bp->fbp_mr->fid);
+	if (ret == FI_SUCCESS) {
+		m0_free(bp);
+		nb->nb_xprt_private = NULL;
+	}
```

Comment:
We need to assert(ret == FI_SUCCESS)?

Otherwise, if failure is returned, the 'bp' is not free-ed. This will cause memory leak.